### ISSUE TITLE
Nested changelog property calls do not work in case of repetition/looping

### ISSUE THREAD

**Issue by zederisch** [liquibase/liquibase#2296]
**Environment**
Liquibase Version: first in 4.4.1 (up to 4.6.2)
Database Vendor & Version: Microsoft SQL Server 15.0.2080
Operating System Type & Version: Windows 10 Enterprise 21H1

**Description**
When defining changelog properties (base values) on a master level and accessing/modifying those again in a nested changelog files (to derive a slightly modified version) and repeating this, the first iteration still works properly, while all following ones fail (nothing happens, no SQL code is generated anymore). In 4.4.0 the attached minimal example works properly, since 4.4.1 no longer.

_**Example:**_
a) Set base value.
- property:
    name: BASE_VALUE
    value: 101
    global: false
b) Modify/extend base value on a deeper level (nested):
- property:
    name: FINAL_VALUE
    value: ${BASE_VALUE}3
    global: false
c) Use final value inside a template configuration file -> Works.
d) Set base value for another instance.
- property:
    name: BASE_VALUE
    value: 102
    global: false
b) Modify/extend base value on a deeper level (nested):
- property:
    name: FINAL_VALUE
    value: ${BASE_VALUE}3
    global: false
c) Use final value inside a template configuration file -> Does not work anymore.

**Steps to reproduce**
You can find the used test setup here:
[property_test.zip](https://github.com/liquibase/liquibase/files/7770128/property_test.zip)

**Comment by nvoxland** [comment]
This will probably be a non-trivial thing to change, so I'm moving it off the quick-fix board for now.

**Comment by zederisch** [comment]
As we honestly would like to upgrade from our very old Liquibase 4.3.5, I tested once again with Liquibase 4.16.1 (currently most recent). Unfortunately, the behavior is still the same and this kills us from moving forward... So any news on this ticket?

I think in the meantime I have a better understanding of the issue, so let me summarize again:

- In Liquibase 4.4.0 and earlier, properties with the "global: false" flag were inherited to nested databaseChangeLogs.
- That means you could define them for an entity on a main level and then call a very general sub-procedure in another configuration file that makes use of them.
- When setting them again within your project flow, they were just overwritten (in contrast to the "global: true" setting, where the first occurance of the property definition wins for ever).
- Since Liquibase 4.4.1, this behavior changed.
- Only properties with the setting "global: true" are at all inherited to sub-procedures. "global: false" only works within the exact same change type.
- But now I cannot overwrite these properties anymore, when I go into the next iteration with another entity, that has in fact different property values, which must be inherited to the identical general sub-procedure doing the job of insert/update.
- As we are using one of the properteis (the primary key of the entity) within the Liquibase ID, also this does not change. As a consequence, only the first appearance of an entity has code generated. Because the primary key property is not updated anymore, all following entities are just not happening at all.

I can change all my so far non-global properties to be now global, so that they are inherited. But then, I would need a mechanism that also global properties are overwritten, when they are set again with a different value within the same SQL code compilation run. The "current appearance" should win, not alwas the first one.

Do you see any chance that this is changed again? Do I have another possibility to achieve my goal, i.e. using isnert/update templates with property values, that change when the same template is called multiple times with different property values. I cannot name the properties differently, as their might be thousands of entities accessing the same template.

Thanks for your help. :)

**Comment by nvoxland** [comment]
I see what you are doing there, @zederisch . Thanks for the better info. It's a fun and cleaver way to be able to reuse changelog files across tenants. 

I don't see this specific behavior being changed back, because it was bugfixes against the defined/described behavior where the bugs were causing problems for people. Unfortunately they were load-bearing bugs for you...

That being said, with your use case I see the value of having options to be able to have perhaps another scope besides just "global or not", or a way to say "yes, override this value" or something. It'll take a little thought on what would work well with the existing "properties are just defined once and not updatable" model without opening up surprising edge cases. I'll open a new ticket for that more general feature request.

In the mean time, the general setup is more "run liquibase multiple times, with the properties passed in differently each time". For example, instead of having your 900-tenant-onboarding.yml with 1001_Tenant1.yml and 1002_Tenant1.yml each setting the property and calling 000-rsw-tenant.yml, you'd make two liquibase calls each using a different USERGROUPID_PREFIX_TENANT_ONBOARDING value passed at runtime as a `-D` CLI argument or env variable or whatever.

Depending on your setup, there may be a way to better script that to still be an overall single call for you?

**Comment by legoethals** [comment]
@zederisch do you have a workaround for this issue? We're using kind of the same strategy, and this is now blocking us from upgrading to Spring boot 3

**Comment by zederisch** [comment]
> I created #3345 to track my proposed feature. Would that address your need?

Hi @nvoxland. First of all, huge apologize for not answering for such a long time... The mail was always in my inbox, but business did not really allow me to dig again into the topic. The new reply in this thread now reminded me again that I still owe you an answer.

So, for me the new "nested" option for the "scope" described in #3345 sounds perfect. I would be very happy to see this. :)

**Comment by zederisch** [comment]
> @zederisch do you have a workaround for this issue? We're using kind of the same strategy, and this is now blocking us from upgrading to Spring boot 3

As mentioned in my comment before, the last half year there was not really enough time to change anything on our Liquibase setup. Therefore, we are actually still using the massively outdated version 4.3.5.

It is always a very big investment for us to change the way our "model" works. The model itself is here not the issue, that's easily changeable, but in case the customer specific entity files calling our general model files also require changes it's getting critical. The database configuration for only one system can easily have hundreds of customer specific configuration files, accessing those model files (i.e. Liquibase templates taking care that certain tables are properly filled). We serve dozenz of different end customers with our model to configure our application, so bigger changes must be always well thought before starting them...

Actually, the full Liquibase approach was always meant as a workaround, because our own application was just missing a possibility for import/export of configuration or staging. Unfortunately, also years later this is now not fully available, but we're getting closer to it. Anyhow, my feeling is the current Liquibase approach has a lot of adventages with respect to collaboration, because every entity of configuration has in the end a single Liquibase configuration file that can be easily maintained in a GIT repository, which allows for merging, cherry picking, etc.

So, maybe the Liquibase model also stays in place even when our own application's staging/import-export functionality is fully in place. I would be definitely interested in upgrading to the most recent version, in case the feature in #3345 comes and nothing else turns out to block us from upgrading. :)

**Comment by dwarcake** [comment]
Tagging along here, using a setup similar as described above.

https://github.com/liquibase/liquibase/issues/3345 Would indeed be a nice solution to the problem.
Currently postponing upgrading to latest Liquibase version
